openstudiocore/resources/model/OpenStudio.idd

```diff
@@ -26411,3 +26411,367 @@ OS:Meter:CustomDecrement,
+       \required-field
+       \type alpha
+       
+OS:EnergyManagementSystem:Actuator,
+       \memo Hardware portion of EMS used to set up actuators in the model
+       \min-fields 4
+  A1, \field Handle
+       \type handle
+       \required-field 
+  A2, \field Name
+       \required-field
+       \type alpha
+       \note This name becomes a variable for use in Erl programs
+       \note No spaces allowed in names for ErlVariableNames
+       \reference ErlVariableNames
+       \reference AllObjects
+  A3, \field Actuated Component Name
+       \required-field
+       \type object-list
+       \object-list AllObjects
+  A4, \field Actuated Component Type
+       \required-field
+       \type alpha
+  A5; \field Actuated Component Control Type
+       \required-field
+       \type alpha      
+
+OS:EnergyManagementSystem:ProgramCallingManager,
+       \extensible:1 - repeat last field, remembering to remove ; from "inner" fields.
+       \memo Input EMS program. a program needs a name
+       \memo a description of when it should be called
+       \memo and then lines of program code for EMS Runtime language
+       \min-fields 4
+  A1, \field Handle
+       \type handle
+       \required-field      
+  A2, \field Name
+       \required-field
+       \type alpha
+       \reference ErlProgramCallingManagerNames
+       \note No spaces allowed in names for ErlProgramCallingManagerNames
+  A3, \field EnergyPlus Model Calling Point
+       \type choice
+       \key BeginNewEnvironment
+       \key AfterNewEnvironmentWarmUpIsComplete
+       \key BeginTimestepBeforePredictor
+       \key AfterPredictorBeforeHVACManagers
+       \key AfterPredictorAfterHVACManagers
+       \key InsideHVACSystemIterationLoop
+       \key EndOfZoneTimestepBeforeZoneReporting
+       \key EndOfZoneTimestepAfterZoneReporting
+       \key EndOfSystemTimestepBeforeHVACReporting
+       \key EndOfSystemTimestepAfterHVACReporting
+       \key EndOfZoneSizing
+       \key EndOfSystemSizing
+       \key AfterComponentInputReadIn
+       \key UserDefinedComponentModel
+       \key UnitarySystemSizing
+  A4; \field Program Name
+       \begin-extensible
+       \required-field
+       \type object-list
+       \note no spaces allowed in name
+       \object-list ErlProgramNames
+
+OS:EnergyManagementSystem:Program,
+       \extensible:1 - repeat last field, remembering to remove ; from "inner" fields.
+       \memo This input defines an Erl program
+       \memo Each field after the name is a line of EMS Runtime Language
+       \min-fields 3
+  A1, \field Handle
+       \type handle
+       \required-field       
+  A2, \field Name
+       \required-field
+       \type alpha
+       \reference ErlProgramNames
+       \reference ErlProgramSubroutineNames
+       \note No spaces allowed in names for ErlProgramNames
+  A3; \field Program Line
+       \begin-extensible
+       \type alpha
+       \required-field
+
+OS:EnergyManagementSystem:Subroutine,
```

Comment:
yeah, i just went for a 1 to 1 mapping.  Subroutines can only be called by other Programs (and not by ProgramCallingManager)